### ISSUE TITLE
3.29 - navigation systematically fails on existing app

### ISSUE THREAD

**Issue by lorenzomilesi** [software-mansion/react-native-screens#1992]
### Description

I updated from 3.28 to 3.29 and suddenly opening a screen stopped working. In my case I have a visible react-native modal, which contains a button. Once pressed, the modal closes and a new screen opens via navigation.navigate(). After the update, navigation no longer occurs and the app is in an inconsistent state, so it thinks it is in a screen that has not been pushed and therefore it is not possible to do anything else.

### Steps to reproduce

1. Build on iOS
2. Open a react-native Modal
3. Close the modal and navigate to a screen

In the Snack below I was not able to add react-native-screens 3.29.0 as not supported from expo sdk 49.0.0

### Snack or a link to a repository

https://snack.expo.dev/@milore/rns-navigation-issue

### Screens version

3.29.0

### React Native version

0.72.8

### Platforms

iOS

### JavaScript runtime

Hermes

### Workflow

React Native (without Expo)

### Architecture

None

### Build type

Release mode

### Device

Real device

### Device model

iPhone 12

### Acknowledgements

Yes

**Comment by kkafar** [comment]
Hey @lorenzomilesi, thanks for reporting!

I can't reproduce the behaviour you describe, after I've fixed the snack you provided. 
Please note that in your snack you navigate nowhere: you create arrow function object (in `onPress` callback) and don't even execute it. See snippet below.

<details>
<summary>Fixed snack</summary>

```typescript
import React from 'react';
import { createNativeStackNavigator, NativeStackNavigationProp } from '@react-navigation/native-stack';
import { NavigationContainer, ParamListBase, useNavigation } from '@react-navigation/native';
import { Button, Modal, View, Text, TouchableOpacity, StyleSheet } from 'react-native';

const Stack = createNativeStackNavigator();

type Props = {
  navigation: NativeStackNavigationProp<ParamListBase>;
};

function Component({ navigation }: Props) {
  const [visible, setVisible] = React.useState(false);
  // const navigation = useNavigation();

  const onPress = () => {
    console.log('onPress clicked');
    setVisible(false);         // <<<============== FIX
    navigation.navigate('Screen2');
  };

  return (
    <View style={{ flex: 1 }}>
      <Button title="hihihi" onPress={() => setVisible(true)} />
      <Modal visible={visible} transparent={true} animationType="fade">
        <View style={[styles.redbox, styles.centeredView]}>
          <TouchableOpacity onPress={onPress}>
            <Text>Go to screen2</Text>
          </TouchableOpacity>
        </View>
      </Modal>
    </View>
  );
}

function Component2({ navigation }: Props) {
  const [visible, setVisible] = React.useState(false);

  const onPress = () => {
    console.log('onPress clicked');
    setVisible(false);
    navigation.navigate('Screen1');
  };

  return (
    <View style={{ flex: 1 }}>
      <Button title="hihihi" onPress={() => setVisible(true)} />
      <Modal visible={visible} transparent={true} animationType="fade">
        <View style={[styles.greenbox, styles.centeredView]}>
          <TouchableOpacity onPress={onPress}>
            <Text>Go to screen1</Text>
          </TouchableOpacity>
        </View>
      </Modal>
    </View>
  );
}

const App = () => (
  <NavigationContainer>
    <Stack.Navigator>
      <Stack.Screen name="Screen1" component={Component} />
      <Stack.Screen name="Screen2" component={Component2} />
    </Stack.Navigator>
  </NavigationContainer>
);

const styles = StyleSheet.create({
  redbox: {
    backgroundColor: 'red',
  },
  greenbox: {
    backgroundColor: 'green',
  },
  centeredView: {
    flex: 1,
    alignContent: 'center',
    justifyContent: 'center',
  },
});

export default App;
```
</details>

After setting the state properly it seems to be working fine, see recording below:

https://github.com/software-mansion/react-native-screens/assets/50801299/810235dc-7235-4baa-bb13-a6dee3db4bff


Please provide us with working reproduction if missing call to `setVisible` was only a typo & the issue lies somewhere else. I'll reopen the issue then.

**Comment by lorenzomilesi** [comment]
Hi @kkafar , idk what happened with that snack, but I think I didn't properly save the last changes. I'm sorry for the extra time you spent.

Have you tried with version 3.29? The problem is there and in no other version.
If no, I will find a moment the next days to create a reproducible repo example.

**Comment by kkafar** [comment]
> Have you tried with version 3.29? The problem is there and in no other version

Hi, yeah, I'm always testing issues on current `main` & I doubt it is something that was fixed after 3.29 was released, but who knows ;D 

One thing I can say for sure: the scenario you described works for me as presented in the recording above (seems correct).

Tag me once you have a reproduction 😃

**Comment by lorenzomilesi** [comment]
Turned out my problem is related to [this issue](https://github.com/react-navigation/react-navigation/issues/5458). Precisely when navigating from a react native `<Modal />` to a React Navigation screen with `presentation: 'modal'`. Some changes in 3.29 highlighted the problem. I will go for a custom implementation of modals.
